File: check_rakuten_32byte.py
# -*- coding: utf-8 -*-
"""
楽天CSVの「バリエーション1選択肢定義」を検証する。
楽天の仕様: 「各選択肢」（|で区切った1つ1つ）が32バイトまで。列全体ではない。
コード.js の truncateToByteLength と同じ数え方: charCode < 128 → 1バイト、それ以外 → 2バイト
"""
import csv
import os
import sys

COL_NAME = 'バリエーション1選択肢定義'
MAX_BYTES = 32

def byte_length(s):
    if s is None or s == '':
        return 0
    s = str(s).strip()
    n = 0
    for c in s:
        n += 1 if ord(c) < 128 else 2
    return n

def check_row(val):
    """|で区切った各選択肢が32バイト以下か。超過した選択肢があれば (選択肢の表示用, バイト数) のリストで返す"""
    if not val or not str(val).strip():
        return []
    over = []
    for part in str(val).split('|'):
        seg = part.strip()
        if not seg:
            continue
        blen = byte_length(seg)
        if blen > MAX_BYTES:
            over.append((seg[:30] + ('…' if len(seg) > 30 else ''), blen))
    return over

def main():
    csv_dir = os.path.join(os.path.dirname(__file__), '楽天CSVファイル')
    if not os.path.isdir(csv_dir):
        print('楽天CSVファイル フォルダが見つかりません')
        sys.exit(1)
    files = [f for f in os.listdir(csv_dir) if f.endswith('.csv')]
    files.sort()
    total_rows = 0
    total_over = 0
    results = []
    for fname in files:
        path = os.path.join(csv_dir, fname)
        try:
            with open(path, 'r', encoding='utf-8') as f:
                reader = csv.reader(f)
                header = next(reader, None)
                if not header:
                    results.append((fname, '空', 0, []))
                    continue
                try:
                    idx = header.index(COL_NAME)
                except ValueError:
                    results.append((fname, 'ヘッダーに列なし', 0, []))
                    continue
                over = []
                count = 0
                for row in reader:
                    if len(row) <= idx:
                        continue
                    count += 1
                    val = (row[idx] or '').strip()
                    if not val:
                        continue
                    seg_over = check_row(val)
                    if seg_over:
                        item = (row[0] if row else '') or ('行' + str(count + 1))
                        for disp, blen in seg_over:
                            over.append((count + 1, item, disp, blen))
                total_rows += count
                total_over += len(over)
                results.append((fname, count, over, None))
        except UnicodeDecodeError:
            try:
                with open(path, 'r', encoding='cp932') as f:
                    reader = csv.reader(f)
                    header = next(reader, None)
                    if not header:
                        results.append((fname, '空', 0, []))
                        continue
                    try:
                        idx = header.index(COL_NAME)
                    except ValueError:
                        results.append((fname, 'ヘッダーに列なし', 0, []))
                        continue
                    over = []
                    count = 0
                    for row in reader:
                        if len(row) <= idx:
                            continue
                        count += 1
                        val = (row[idx] or '').strip()
                        if not val:
                            continue
                        seg_over = check_row(val)
                        if seg_over:
                            item = (row[0] if row else '') or ('行' + str(count + 1))
                            for disp, blen in seg_over:
                                over.append((count + 1, item, disp, blen))
                    total_rows += count
                    total_over += len(over)
                    results.append((fname, count, over, None))
            except Exception as e:
                results.append((fname, str(e), 0, []))
        except Exception as e:
            results.append((fname, str(e), 0, []))
    # 出力
    print('=== 楽天「バリエーション1選択肢定義」32バイト超過チェック（各選択肢ごと）===')
    print('楽天仕様: |で区切った「各選択肢」が32バイトまで。数え方: ASCII=1バイト、それ以外=2バイト\n')
    for r in results:
        if len(r) == 4 and r[3] is None:
            fname, data_rows, over, _ = r
            print('ファイル:', fname)
            print('  データ行数:', data_rows, '  32バイト超過:', len(over))
            if over:
                for row_num, item, val, blen in over[:15]:
                    print('    行{}  {}  バイト数={}  値="{}"'.format(row_num, item, blen, val))
                if len(over) > 15:
                    print('    ... 他', len(over) - 15, '件')
            print()
        else:
            print('ファイル:', r[0], '  ', r[1])
            print()
    print('--- まとめ ---')
    print('対象データ行合計:', total_rows)
    print('32バイト超過した選択肢の数:', total_over)
    if total_over == 0 and total_rows > 0:
        print('→ すべての選択肢が32バイト以下でした。')
    elif total_over > 0:
        print('→ 上記の選択肢が楽天の32バイト制限を超えています。')

if __name__ == '__main__':
    main()

File: check-rakuten-32byte.js
/**
 * 楽天CSVの「バリエーション1選択肢定義」が32バイト以下か検証する。
 * コード.js の truncateToByteLength と同じ数え方: charCode < 128 → 1バイト、それ以外 → 2バイト
 * 使い方: node check-rakuten-32byte.js
 */
const fs = require('fs');
const path = require('path');

function byteLength(str) {
  if (str == null || str === '') return 0;
  const s = String(str);
  let bytes = 0;
  for (let i = 0; i < s.length; i++) {
    bytes += s.charCodeAt(i) < 128 ? 1 : 2;
  }
  return bytes;
}

function parseCsvLine(line) {
  const fields = [];
  let cur = '';
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (c === '"') {
      inQuotes = !inQuotes;
    } else if ((c === ',' && !inQuotes) || (c === '\n' && !inQuotes)) {
      fields.push(cur);
      cur = '';
      if (c === '\n') break;
    } else {
      cur += c;
    }
  }
  if (cur !== '' || line[line.length - 1] === ',') fields.push(cur);
  return fields;
}

const csvDir = path.join(__dirname, '楽天CSVファイル');
const logDir = path.join(__dirname, '楽天ログファイル');
const COL_NAME = 'バリエーション1選択肢定義';
const MAX_BYTES = 32;

const filesToCheck = [];
if (fs.existsSync(csvDir)) {
  const mainCsv = path.join(csvDir, 'normal-item.csv');
  if (fs.existsSync(mainCsv)) filesToCheck.push({ path: mainCsv, name: '楽天CSVファイル/normal-item.csv' });
}
if (fs.existsSync(logDir)) {
  const names = fs.readdirSync(logDir).filter(n => n.endsWith('.csv'));
  names.forEach(n => filesToCheck.push({ path: path.join(logDir, n), name: '楽天ログファイル/' + n }));
}

if (filesToCheck.length === 0) {
  console.log('CSVファイルが見つかりません。');
  process.exit(1);
}

let totalOver = 0;
let totalRows = 0;
const results = [];

for (const { path: filePath, name } of filesToCheck) {
  let content;
  try {
    content = fs.readFileSync(filePath, 'utf8');
  } catch (e) {
    try {
      content = fs.readFileSync(filePath, 'shiftjis').toString();
    } catch (e2) {
      console.warn('スキップ（読めません）:', name);
      continue;
    }
  }
  const lines = content.split(/\r?\n/).filter(l => l.trim() !== '');
  if (lines.length < 2) continue;
  const headerFields = parseCsvLine(lines[0]);
  const idx = headerFields.findIndex(h => h.trim() === COL_NAME);
  if (idx === -1) {
    results.push({ file: name, note: 'ヘッダーに「バリエーション1選択肢定義」なし', over: [], overCount: 0, dataRows: 0 });
    continue;
  }
  const over = [];
  let dataRows = 0;
  for (let r = 1; r < lines.length; r++) {
    const fields = parseCsvLine(lines[r]);
    if (fields.length <= idx) continue;
    dataRows++;
    const val = (fields[idx] || '').trim();
    if (val === '') continue;
    const bytes = byteLength(val);
    if (bytes > MAX_BYTES) {
      const itemUrl = fields[0] || ('行' + (r + 1));
      over.push({ row: r + 1, itemUrl, value: val.substring(0, 50) + (val.length > 50 ? '…' : ''), bytes });
    }
  }
  totalRows += dataRows;
  totalOver += over.length;
  results.push({ file: name, over, overCount: over.length, dataRows });
}

console.log('=== 楽天「バリエーション1選択肢定義」32バイト超過チェック ===\n');
console.log('数え方: コード.js の truncateToByteLength と同じ（ASCII=1バイト、それ以外=2バイト）\n');

for (const r of results) {
  console.log('ファイル:', r.file);
  if (r.note) {
    console.log('  ', r.note);
    continue;
  }
  console.log('  データ行数:', r.dataRows, '  32バイト超過:', r.overCount);
  if (r.over.length > 0) {
    r.over.forEach(o => {
      console.log('    行' + o.row + '  ' + o.itemUrl + '  バイト数=' + o.bytes + '  値="' + o.value + '"');
    });
  }
  console.log('');
}

console.log('--- まとめ ---');
console.log('対象データ行合計:', totalRows);
console.log('32バイト超過の行数:', totalOver);
if (totalOver === 0 && totalRows > 0) {
  console.log('→ すべて32バイト以下でした。');
} else if (totalOver > 0) {
  console.log('→ 上記の行が楽天の32バイト制限を超えています。');
}
